Speculation Rules: Fix waitForResponse test scenarios

yoavweiss opened the pull request:
#### fed984d0feb9b3773d13dfa6270fde810f980994
<pre>
Speculation Rules: Fix waitForResponse test scenarios
<a href="https://bugs.webkit.org/show_bug.cgi?id=300582">https://bugs.webkit.org/show_bug.cgi?id=300582</a>

Reviewed by Alex Christensen.

The navigation-timing waitForResponse and waitForRedirect tests are testing two different things:
(1) That prefetch is working correctly when the prefetched response starts arriving after the navigation to that response.
(2) That when that happens, the navigation timing values for responseStart and requestStart are correct.

Fixing (1) required avoiding canceling loaders for prefetch resources when navigating.
Fixing (2) required making sure that the navigation timing values cannot be negative (relative to the current document&apos;s timeOrigin).

No new tests, but this adjusts existing test expectations.

* LayoutTests/TestExpectations: Remove skipping of the navigation-timing tests.
* LayoutTests/imported/w3c/web-platform-tests/speculation-rules/prefetch/navigation-timing-requestStart-responseStart.https_include=waitingForRedirect-expected.txt: Progression.
* LayoutTests/imported/w3c/web-platform-tests/speculation-rules/prefetch/navigation-timing-requestStart-responseStart.https_include=waitingForResponse-expected.txt: Progression.
* Source/WebCore/loader/DocumentLoader.cpp:
(WebCore::DocumentLoader::stopLoadingSubresources): Don&apos;t stop prefetch resource loaders.
(WebCore::DocumentLoader::loadMainResource): Rename to AllowCachingMainResourcePrefetch.
(WebCore::DocumentLoader::cancelMainResourceLoad): Don&apos;t cancel a main resource prefetch.
* Source/WebCore/loader/DocumentPrefetcher.cpp:
(WebCore::DocumentPrefetcher::prefetch): Rename to AllowCachingMainResourcePrefetch.
* Source/WebCore/loader/ResourceLoaderOptions.h:
* Source/WebCore/loader/cache/CachedResource.cpp:
(WebCore::CachedResource::CachedResource): Rename to AllowCachingMainResourcePrefetch.
(WebCore::CachedResource::addClientToSet): Rename to AllowCachingMainResourcePrefetch.
* Source/WebCore/loader/cache/CachedResource.h:
(WebCore::CachedResource::allowsCaching const):
* Source/WebCore/loader/cache/CachedResourceLoader.cpp:
(WebCore::CachedResourceLoader::shouldUpdateCachedResourceWithCurrentRequest): Rename to AllowCachingMainResourcePrefetch.
(WebCore::isResourceSuitableForDirectReuse): Rename to AllowCachingMainResourcePrefetch.
(WebCore::CachedResourceLoader::updateCachedResourceWithCurrentRequest): Rename to AllowCachingMainResourcePrefetch.
(WebCore::mustReloadFromServiceWorkerOptions): Rename to AllowCachingMainResourcePrefetch.
(WebCore::CachedResourceLoader::determineRevalidationPolicy const): Rename to AllowCachingMainResourcePrefetch.
* Source/WebCore/loader/cache/CachedResourceRequest.h:
(WebCore::CachedResourceRequest::allowsCaching const): Rename to AllowCachingMainResourcePrefetch.
* Source/WebCore/page/PerformanceResourceTiming.cpp:
(WebCore::networkLoadTimeToDOMHighResTimeStamp): Return 0.0 instead of a negative timestamp.
* Source/WebCore/platform/network/CacheValidation.cpp:
(WebCore::updateRedirectChainStatus): Rename to AllowCachingMainResourcePrefetch.

Canonical link: <a href="https://commits.webkit.org/301456@main">https://commits.webkit.org/301456@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6c7f5415b2b9e7f9047df7041cd529a1f932264a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/125988 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/45645 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/36417 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/132837 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/77828 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/127859 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/46321 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/54196 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/95972 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/64066 "") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/128936 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/37052 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/112672 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/76461 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/35960 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/30857 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/76308 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/106834 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/31072 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/135529 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/52760 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/40491 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/104454 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/53212 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/108886 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/104174 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26545 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/49554 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/27882 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/50138 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/52654 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/58476 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/51996 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/55344 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/53694 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebCore/page/PerformanceResourceTiming.cpp:
```diff
@@ -47,7 +47,7 @@ namespace WebCore {
 {
     ASSERT(timeOrigin);
-    if (!timeStamp)
+    if (!timeStamp || timeStamp < timeOrigin)
```
This could be just one check timeStamp <= timeOrigin

yoavweiss replied:
Done

achristensen07 commented on Source/WebCore/loader/DocumentLoader.cpp:
```diff
@@ -1980,7 +1980,16 @@ void DocumentLoader::stopLoadingPlugIns()
 {
-    cancelAll(m_subresourceLoaders);
+    for (auto& loader : copyToVector(m_subresourceLoaders)) {
+        if (auto* cachedResource = loader->cachedResource()) {
```
CachedResourceHandle instead of auto*

yoavweiss replied:
Done

achristensen07 commented on Source/WebCore/loader/DocumentLoader.cpp:
```diff
@@ -2353,6 +2362,10 @@ void DocumentLoader::cancelMainResourceLoad(const ResourceError& resourceError,
     cancelPolicyCheckIfNeeded();
 
+    // Don't cancel loaders for prefetch resources, as they need to survive navigation.
+    if (m_mainResource && m_mainResource->options().cachingPolicy == CachingPolicy::AllowCachingPrefetch)
```
Do you think it would make the code more clear to rename AllowCachingPrefetch to AllowCachingMainResourcePrefetch?

yoavweiss replied:
Renamed

yoavweiss committed:
Speculation Rules: Fix waitForResponse test scenarios
https://bugs.webkit.org/show_bug.cgi?id=300582

Reviewed by Alex Christensen.

The navigation-timing waitForResponse and waitForRedirect tests are testing two different things:
(1) That prefetch is working correctly when the prefetched response starts arriving after the navigation to that response.
(2) That when that happens, the navigation timing values for responseStart and requestStart are correct.

Fixing (1) required avoiding canceling loaders for prefetch resources when navigating.
Fixing (2) required making sure that the navigation timing values cannot be negative (relative to the current document's timeOrigin).

No new tests, but this adjusts existing test expectations.

* LayoutTests/TestExpectations: Remove skipping of the navigation-timing tests.
* LayoutTests/imported/w3c/web-platform-tests/speculation-rules/prefetch/navigation-timing-requestStart-responseStart.https_include=waitingForRedirect-expected.txt: Progression.
* LayoutTests/imported/w3c/web-platform-tests/speculation-rules/prefetch/navigation-timing-requestStart-responseStart.https_include=waitingForResponse-expected.txt: Progression.
* Source/WebCore/loader/DocumentLoader.cpp:
(WebCore::DocumentLoader::stopLoadingSubresources): Don't stop prefetch resource loaders.
(WebCore::DocumentLoader::loadMainResource): Rename to AllowCachingMainResourcePrefetch.
(WebCore::DocumentLoader::cancelMainResourceLoad): Don't cancel a main resource prefetch.
* Source/WebCore/loader/DocumentPrefetcher.cpp:
(WebCore::DocumentPrefetcher::prefetch): Rename to AllowCachingMainResourcePrefetch.
* Source/WebCore/loader/ResourceLoaderOptions.h:
* Source/WebCore/loader/cache/CachedResource.cpp:
(WebCore::CachedResource::CachedResource): Rename to AllowCachingMainResourcePrefetch.
(WebCore::CachedResource::addClientToSet): Rename to AllowCachingMainResourcePrefetch.
* Source/WebCore/loader/cache/CachedResource.h:
(WebCore::CachedResource::allowsCaching const):
* Source/WebCore/loader/cache/CachedResourceLoader.cpp:
(WebCore::CachedResourceLoader::shouldUpdateCachedResourceWithCurrentRequest): Rename to AllowCachingMainResourcePrefetch.
(WebCore::isResourceSuitableForDirectReuse): Rename to AllowCachingMainResourcePrefetch.
(WebCore::CachedResourceLoader::updateCachedResourceWithCurrentRequest): Rename to AllowCachingMainResourcePrefetch.
(WebCore::mustReloadFromServiceWorkerOptions): Rename to AllowCachingMainResourcePrefetch.
(WebCore::CachedResourceLoader::determineRevalidationPolicy const): Rename to AllowCachingMainResourcePrefetch.
* Source/WebCore/loader/cache/CachedResourceRequest.h:
(WebCore::CachedResourceRequest::allowsCaching const): Rename to AllowCachingMainResourcePrefetch.
* Source/WebCore/page/PerformanceResourceTiming.cpp:
(WebCore::networkLoadTimeToDOMHighResTimeStamp): Return 0.0 instead of a negative timestamp.
* Source/WebCore/platform/network/CacheValidation.cpp:
(WebCore::updateRedirectChainStatus): Rename to AllowCachingMainResourcePrefetch.

Canonical link: https://commits.webkit.org/301456@main
